internal/pkg/service/stream/storage/level/local/writer/writer.go

```diff
@@ -1,42 +1,360 @@
+	CompletedWrites() uint64
+	// FirstRecordAt returns timestamp of receiving the first record for processing.
+	FirstRecordAt() utctime.UTCTime
+	// LastRecordAt returns timestamp of receiving the last record for processing.
+	LastRecordAt() utctime.UTCTime
+	// CompressedSize written to the file, measured after compression writer.
+	CompressedSize() datasize.ByteSize
+	// UncompressedSize written to the file, measured before compression writer.
+	UncompressedSize() datasize.ByteSize
+}
+
+// writer implements Writer interface, it wraps common logic for all file types.
+// For conversion between record values and bytes, the Writer is used.
+type writer struct {
+	dirPath  string
+	filePath string
+
+	logger log.Logger
+	slice  *model.Slice
+	events *Events
+
+	chain  *writechain.Chain
+	syncer *disksync.Syncer
+
+	formatWriter FormatWriter
+	// closed blocks new writes
+	closed chan struct{}
+	// writeWg waits for in-progress writes before Close
+	writeWg *sync.WaitGroup
+
+	acceptedWrites    *count.Counter
+	completedWrites   *count.CounterWithBackup
+	compressedMeter   *size.MeterWithBackup
+	uncompressedMeter *size.MeterWithBackup
+}
+
+func New(
+	ctx context.Context,
+	logger log.Logger,
+	clk clock.Clock,
+	cfg Config,
+	slice *model.Slice,
+	file writechain.File,
+	dirPath string,
+	filePath string,
+	syncerFactory disksync.SyncerFactory,
+	formatWriterFactory FormatWriterFactory,
+	volumeEvents *Events,
+) (out Writer, err error) {
+	w := &writer{
+		dirPath:  dirPath,
+		filePath: filePath,
+		logger:   logger.WithComponent("slice-writer"),
+		slice:    slice,
+		events:   volumeEvents.Clone(), // clone volume events, to attach additional writer specific events
+		closed:   make(chan struct{}),
+		writeWg:  &sync.WaitGroup{},
+	}
+
+	// Close resources on error
+	defer func() {
+		if err != nil {
+			if w.syncer != nil {
+				_ = w.syncer.Stop(ctx)
+			}
+			if w.chain != nil {
+				_ = w.chain.Close(ctx)
+			}
+			if w.completedWrites != nil {
+				_ = w.completedWrites.Close()
+			}
+		}
+	}()
+
+	// Create counters
+	{
+		// In progress writes counter
+		w.acceptedWrites = count.NewCounter()
+
+		// Successful writes counter, the value backup is periodically saved to disk.
+		// In the case of non-graceful node shutdown, the initial state is loaded from the disk after the node is restarted.
+		// In that case, the statistics may not be accurate, but this should not happen, and we prefer throughput over the atomicity of statistics.
+		w.completedWrites, err = count.NewCounterWithBackupFile(ctx, clk, logger, filepath.Join(dirPath, CompletedWritesCounter), cfg.Statistics.DiskSyncInterval)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	// Create empty chain of writers to the file
+	{
+		w.chain = writechain.New(logger, file)
+	}
+
+	// Add a buffer before the file
+	{
+		if cfg.FileBuffer > 0 {
+			w.chain.PrependWriter(func(w io.Writer) io.Writer {
+				return limitbuffer.New(w, int(cfg.FileBuffer.Bytes()))
+			})
+		}
+	}
```

Comment:
Important buffer before the file, ... so when we are waiting for the sync to the file - new CSV rows can be generated, compresed and prepared in the buffer.

Reply:
(chain in composed from the end, from the output file)